File: Leetcode_All_Solved/130.cpp
class Solution {
private:
    int m, n;
    void dfs(vector<vector<char>>& board, vector<vector<bool>>& visited, int i, int j, bool capture=false){
        if(!(0<=i && i<m && 0<=j && j<n))return;
        if(board[i][j] != 'O')return;
        if(visited[i][j] == true)return;
        visited[i][j] = true;
        
        if(capture == true) board[i][j] = 'X';
        
        dfs(board,visited,i+1,j,capture);
        dfs(board,visited,i-1,j,capture);
        dfs(board,visited,i,j+1,capture);
        dfs(board,visited,i,j-1,capture);
    }
public:
    void solve(vector<vector<char>>& board) {
        m = board.size(), n = board[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        for(int i=0;i<m;i++){
            if(board[i][0] == 'O' && visited[i][0]==false)dfs(board, visited, i, 0);
            if(board[i][n-1] == 'O' && visited[i][n-1]==false)dfs(board, visited, i, n-1);
        }
        for(int j=0;j<n;j++){
            if(board[0][j] == 'O' && visited[0][j]==false)dfs(board, visited, 0, j);
            if(board[m-1][j] == 'O' && visited[m-1][j]==false)dfs(board, visited, m-1, j);
        }
        
        //capture
        for(int i=0;i<m;i++)for(int j=0;j<n;j++)if(board[i][j]=='O' && visited[i][j]==false)
            dfs(board, visited, i, j, true);
    }
};

File: Leetcode_All_Solved/130.py
# O(mn)
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m = len(board)
        n = len(board[0])
        visited = [[False for j in range(n)] for i in range(m)]
        di = [1, 0, -1, 0]
        dj = [0, 1, 0, -1]
        

        def dfs(i: int, j: int, capture: bool) -> None:
            if(not(0<=i and i<m and 0<=j and j<n)):
                return

            if(visited[i][j] == True or board[i][j] != "O"):
                return
            visited[i][j] = True

            if(capture == True):
                board[i][j] = "X"
            
            for a, b in zip(di, dj):
                dfs(i+a, j+b, capture)
            return


        def check(i: int, j: int, capture: bool) -> None:
            if(visited[i][j] == False and board[i][j] == "O"):
                dfs(i, j, capture)
        
        for i in range(m):
            check(i, 0, False)
            check(i, n-1, False)
        for j in range(n):
            check(0, j, False)
            check(m-1, j, False)
        
        for i in range(1, m-1):
            for j in range(1, n-1):
                check(i, j, True)
        return
